node/impl/full/state_v2.go

```diff
@@ -0,0 +1,84 @@
+package full
+
+import (
+	"context"
+
+	"go.uber.org/fx"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+var _ StateModuleAPIv2 = (*StateModuleV2)(nil)
+
+type StateModuleAPIv2 interface {
+	StateGetActor(context.Context, address.Address, types.TipSetSelector) (*types.Actor, error)
+	StateGetID(context.Context, address.Address, types.TipSetSelector) (*address.Address, error)
+	StateCompute(context.Context, []*types.Message, types.TipSetSelector) (*api.ComputeStateOutput, error)
+	StateSimulate(context.Context, []*types.Message, types.TipSetSelector, types.TipSetLimit) (*api.ComputeStateOutput, error)
+}
+
+type StateModuleV2 struct {
+	State StateAPI
+	Chain ChainModuleV2
+
+	fx.In
+}
+
+// TODO: Discussion: do we want to optionally strict the type of selectors that
+//       can be supplied here to avoid foot-guns? For example, only accept TipSetKey
```

Comment:
What foot guns are you anticipating?

Reply:
[This](https://www.notion.so/filecoindev/Filecoin-V2-APIs-1d0dc41950c1808b914de5966d501658#1d0dc41950c180dc8eb9d1afc46cc641) specifically, where a naïve programmer would repeatedly use tipset selectors with no concrete key for calls that should care about cross consistency.

Reply:
I want to be able to call this (and other) APIs with `"latest"`, that's pretty typical because you often have good reason to believe that (a) the actor stably exists and (b) you're not expecting to use the return for anything critical. I suppose switching to `"finalized"` as the default usage pattern would be ideal after activation. But if we allow those tags, what else is there to restrict?